standards/core/rln-contract.md

```diff
@@ -0,0 +1,301 @@
+- registration MUST fail if the total rate limit of _Active_, _GracePeriod_, and _Expired_ memberships, including the one being created, would exceed the maximum total rate;
+- registration MUST fail if the requested rate limit for a new membership is lower than the minimal allowed rate limit;
+- the user MUST lock-up a deposit to register a membership;
+- the user MUST specify the rate limit of the new membership[^2];
+- the size of the deposit MUST depend on the requested rate limit;
+- in case of a successful registration:
+	- the new membership MUST become _Active_;
+	- the current total rate limit MUST be incremented by the rate limit of the new membership;
+- a newly created membership MUST have an expiration time `T` and a grace period `G`.
+
+[^2]: A user-facing application SHOULD suggest default values for rate limits for the user.
+
+### Extend a membership
+
+Extending a membership is subject to the following conditions:
+- extension MUST fail if the membership is in any state other than _GracePeriod_;
+- the membership holder MUST be able to extend their membership;
+- any user except the membership holder MUST NOT be able to extend a membership;
+- after a successful extension, the membership MUST become _Active_.
+
+Holding a membership means controlling the private key from which the RLN commitment ID (i.e., public key) was derived.
+
+### Withdraw the deposit
+
+Deposit withdrawal is subject to the following conditions:
+- the membership holder MUST be able to withdraw their deposit;
+- any user except the membership holder MUST NOT be able to withdraw its deposit;
+- a deposit MUST be withdrawn in full;
+- a withdrawal MUST fail if the membership is not in _GracePeriod_, _Expired_, or _ErasedAwaitsWithdrawal_;
+- a membership MUST become _Erased_ after withdrawal.
+
+## Governance and upgradability
+
+At initial mainnet deployment, the contract MUST have an _Owner_.
+The _Owner_ MUST be able to change the values of all contract parameters.
+The _Owner_ MUST be able to pause any of the following contract functionalities:
+- register a membership;
+- extend a membership;
+- withdraw a deposit.
+
+At some point, the _Owner_ SHOULD renounce their privileges, and the contract MUST become immutable.
+Further upgrades, if necessary, SHOULD be done by deploying a new contract and migrating the membership set.
+
+## Implementation Suggestions
+
+User-facing application SHOULD suggest a few rate limits (tiers) to simplify their users' choice.
+The RECOMMENDED rate limits in a three-tier model are as follows:
+- `20` messages per epoch as low-tier;
+- `200` messages per epoch as mid-tier;
+- `600` messages per epoch as high-tier.
+Users SHOULD be able to select a custom rate limit under advanced settings.
+
+The RECOMMENDED pricing parameters are:
+
+| Parameter          | Value  |
+| ------------------ | ------ |
+| Accepted tokens    | `DAI`  |
+| Reference currency | `USD`  |
+| Pricing function   | linear |
+
```

Comment:
In this _Implementation Suggestions_ section, it may be good to separate suggestion to application developers (1st and 3rd) and suggestion smart contract developers/deployers.

Reply:
I thought that both the pricing function and the accepted currency are choices that smart contract deployers make, am i wrong? In particular, the pricing function (the way I see it) is to be encoded either in the main contract, or (for better upgradability) in another contract that is called from the main contract. Same for `DAI` as the pricing currency: is this really up for the app developer to decide? If I, as an app developer, want my app to accept `ETH`, for example, this won't change the fact that the deployed RLN contract accepts `DAI`. I'd need to involve an oracle and convert, sure, but isn't this out of scope for this spec?

Reply:
My point exactly. The "Implementation suggestion" contains suggestion for application developers that use a smart contract already deployed:
1. The tiers to present to users
2. Saving expiration date and notifying.

And it also contain a suggestion for developers that deploy a smart contract: the token to accept.

This is confusing because if you used a SC already deployed, then you have no choice on the token.

Reply:
Gotcha. I'll move the recommendations to smart contract developers / deployers to an earlier parameter table, and the "Implementation suggestions" section would then only contain advice for app developers.